### ISSUE TITLE
Create stores on the fly

### ISSUE THREAD

**Issue by Templarian** [sinclairzx81/smoke-database#1]
We can use this issue to think over a way to handle such a feature or if it's even needed.

So, there is no way to know the version of a indexedDB... unless you store it in indexedDB. 🤔 

Basically what would be best is to define something like...

```typescript
  const database = new smokedb.DynamicDatabase({
      name    : "todo-database",
      stores  : [] // Optional
  })
```

This creates a version `1` database and creates a store `info`. That simply contains a single integer value of starting with 1.

Basically you'll need to hide the ugly-ness of the the code below inside of the `database.store()` call (and I assume modify or update if it already exists).

```typescript
// basically a store call specific to the info table to get the two values in it, name and version
await database.info((name, version) => {
  // Add to array of upgrades... this will always be the newest
  await database.upgrade(version + 1, (db) => {
    // on submit... and then add the indexes
    // db.createObjectStore(storeName, { keyPath: primaryIndex });
  });
});
```

Then it can look clean like.

```typescript
await database.store('foo')
  .createPrimaryIndex('username')
  .createIndex('first_idx', 'first', {unique: false});
  .createIndex('last_idx', 'last', {unique: false})
  .submit()
```

Little drunk still from earlier, but that's a rough idea. 😕

**Comment by sinclairzx81** [comment]
Awesome, thanks for this !

So, had a quick play around with the ```IDBFactory.open()``` method this morning. 

It appears that the ```version``` passed on ```IDBFactory.open()``` is optional. Basically the call to 
```IDBFactory.open()``` will automatically resolve the current version, with the version number itself available on the ```IDBDatabase``` interface. Works for chrome/firefox/opera.

see https://developer.mozilla.org/en-US/docs/Web/API/IDBFactory/open

### version argument
> Optional. The version to open the database with. If the version is not provided and the database exists, then a connection to the database will be opened without changing its version. If the version is not provided and the database does not exist, then it will be created with version 1.

So, i think this could lead to a workable solution. The following are my current thoughts.

#### opening the connection
1) caller opens the database (via Database constructor)
2) smoke inspects the requested database, storing the current version + stores (schema)

note: if the database does not exist, then it is created with empty stores. Can assume this default behaviour. This version automatically resolves to 1.
note: the word "schema" implies the some representation of stores + indices gathered on first load. The schema can be used to track later schema changes issued by the caller.

#### caller requests unknown object store
1) smoke looks up stored schema, and sees store does not exist.
2) smoke closes database.
3) smoke reopens database using ```databaseUpgrade(version + 1)``` method (to be written) that accepts
stored version (with the version gathered on step 2 above). 
4) smoke returns store to caller.

So basically, this solution would lead to two distinct database opening procedures.

1) ```databaseOpen(name)``` - first time open - loads schemas.
2) ```databaseUpgrade(name, version, changes)``` - subsequent open where the caller is requesting operations that require schema changes.

Could work :)

**Comment by sinclairzx81** [comment]
Have opted against creating stores on the fly. Instead opting for users to specify the stores they need upfront. I found this simplifies the versioning logic quite a bit (happens once at start up), and means that requests to stores (the common use case) don't need to involve version checks on each request to said store.

Version changes are now resolved within the Database constructor, and there is logic there to automatically add and remove stores based on what the constructor in its store list. For example..

```typescript
// first page load:
// creates store1 and store2.
const database = new Database("app", ["store1", "store2"]) 

// second page load:
// store1 retained, store2 dropped, store3 added.
const database = new Database("app", ["store1", "store3"]) 
```
There is however no indexing being applied to these stores (other that the key), as there is nothing to tie the indexes to with the queries themselves. This is still something worth rethinking.

Cheers
